zend_abstract_interface/functions/functions.h

```diff
@@ -0,0 +1,32 @@
+#ifndef ZAI_FUNCTIONS_H
+#define ZAI_FUNCTIONS_H
+
+#include <main/php.h>
+#include <stdbool.h>
+
+/* Work in progress
+ *
+ * The long-term goal is to provide ZAI data-structure shims so that the
+ * following APIs are the same across all PHP versions. For now we will provide
+ * ZAI seams using native Zend Engine data structures.
+ */
+
+/* Calls a function with 'argc' number of zval arguments. Caller owns the
+ * arguments and the 'retval'. Caller is responsible for freeing any refcounted
+ * arguments. Caller must provide a pointer to a stack-allocated zval for the
+ * return value, 'retval'. In error cases, 'retval' will be set to IS_UNDEF.
+ * Caller must dtor the 'retval' after the call:
```

Comment:
I'm not sure I understand what you're suggesting in terms of the refcount handling, but we certainly can (and will 🙂) add an abstraction layer/wrapper at the tracer level. (For example, auto-flushing on PHP 5 wraps the method-call seam with an [API that ignores the retval](https://github.com/DataDog/dd-trace-php/blob/5275a4f/ext/php5/auto_flush.c#L9-L17).)

On the refcounting bit - I think the comment might be a bit confusing the way it is worded.

- Basically the caller 100% owns the args (ctoring/passing/dtoring).
- The caller also owns the space for the retval (ideally stack-allocated).
- The only memory `zai_call_function` allocates is the space to hold the params to pass to the function and it is freed before returning to the caller. (As a side note we might want to consider a hard-limit on the number of args so that we can stack-allocate the params. 🤔)
- The retval may or may not need to be dtored depending on whether or not the PHP function returns a refcounted value (e.g. array, object, string, etc). But that should be known to the caller so the caller can make that choice what to do with the retval.

Reply:
> [W]e might want to consider a hard-limit on the number of args so that we can stack-allocate the params. 🤔

Allowing the caller to manage the memory so it can stack allocate with exactly the right number without any limits seems like it might be better. I think it may lose some expressiveness, though? Is that why you went this way?

Reply:
I'm not sure why we need a hard limit for stack allocated arrays? simple variable length array: `zval params[argc];`. (At least I assume that we can use C99 here?)

Reply:
> I'm not sure why we need a hard limit for stack allocated arrays? simple variable length array: zval params[argc];. (At least I assume that we can use C99 here?)

Indeed we can; we use C99 and C11. Thanks for the tip. I've refactored the param handling in 8ae546bf.

Reply:
For what it is worth, C11 moved VLA's to a conditional feature instead of mandatory. The compiler community seems to want to get rid of this feature. I think I still might prefer moving it to the caller, where the size is known.

Reply:
> I think I still might prefer moving it to the caller, where the size is known.

Do you see a significant advantage for moving this to the caller vs using a variable length array?

Although this might save us some assignments, my main concern is that the API will not be as expressive which would make it easier pass the params the wrong way. Since one of the main purposes of the ZAI API is to increase stability by reducing the chance of using the API incorrectly, my preference would be to keep it as is and reduce the cognitive load of using the API.

Reply:
Although, [UBSan does not like](https://app.circleci.com/pipelines/github/DataDog/dd-trace-php/5787/workflows/560dc271-42ab-4d60-8816-d46b0e29260e/jobs/435966/steps) our VLA with zero length (calling a function without args).

```
/home/circleci/datadog/zend_abstract_interface/functions/php8/functions.c:68:10: runtime error: variable length array bound evaluates to non-positive value 0
```

I'll need to fix this tomorrow. But 🎉  for having the `*San`s catch this in CI!

Reply:
Thanks for the clarification! :)

Reply:
I've addressed the UBSan errors in e8392df. (Marking this unresolved until hearing thoughts on moving args to the caller from @morrisonlevi.)

Reply:
If it weren't for UX and minimizing the space for users to get things wrong, I would prefer caller allocated memory over variable length arrays as compiler vendors seem to really dislike it. Notably, MSVC doesn't support VLAs (or if it does it was only recently changed). I think having an upper-bound on argument length as previously suggested would be a decent compromise, as I think we should not intentionally add any more friction to eventually supporting Windows.

Reply:
Thanks for sharing the info. I've updated the args to use a small fixed size array in 5d3566eb.